lib/semantics/tools.cc

```diff
@@ -442,6 +488,31 @@ bool IsAssumedSizeArray(const Symbol &symbol) {
   return details && details->IsAssumedSize();
 }
 
+bool IsExternalInPureContext(const Symbol &symbol, const Scope &scope) {
+  if (const auto *pureProc{semantics::FindPureProcedureContaining(&scope)}) {
+    if (const Symbol * root{GetAssociationRoot(symbol)}) {
+      if (semantics::FindExternallyVisibleObject(*root, *pureProc)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool InProtectedContext(const Symbol &symbol, const Scope &currentScope) {
+  return IsProtected(symbol) && !IsHostAssociated(symbol, currentScope);
+}
+
+bool IsModifiable(const Symbol &symbol, const Scope &scope) {
+  if (const Symbol * root{GetAssociationRoot(symbol)}) {
+    return IsVariableName(*root) && !InProtectedContext(*root, scope) &&
+        !IsExternalInPureContext(*root, scope) &&
+        !IsOrContainsEventOrLockComponent(*root);
```

Comment:
Here we know why the symbol can't appear in a variable definition context, so it would be nice to share that with the user. Presumably we will be doing the same check in lots of places.

Maybe return an optional message that says what's wrong with symbol, like "%s is not a variable". The main error message would be "%s may not appear in a locality spec" and then attach the returned message to it.

Reply:
Will do.